### PR TITLE
Add document-oriented completion suggester

### LINKED ISSUE #10746: Completion Suggester V2
# Completion Suggester V2

The completion suggester provides auto-complete/search-as-you-type functionality. 
This is a navigational feature to guide users to relevant results as they are typing, improving search precision. 
It is not meant for spell correction or did-you-mean functionality like the term or phrase suggesters.

The completions are indexed as a weighted FST (finite state transducer) to provide fast Top N prefix-based
searches suitable for serving relevant results as a user types.

**Notable Features**: 
- Document oriented suggestions:
 - Near-real time.
 - Deleted document filtering.
 - Multiple Context support.
 - Return document field values via `payload`. 
- Query Interface:
 - Regular expression support via `regex`.
 - Typo tolerance via `fuzzy`.
 - Context boosting at query time.

Completion Suggester V2 is based on [LUCENE-6339] (https://issues.apache.org/jira/browse/LUCENE-6339) and [LUCENE-6459] (https://issues.apache.org/jira/browse/LUCENE-6459), the first iteration of Lucene's new *suggest* API.

## Mapping
The completion fields are indexed in a special way, hence a field mapping has to be defined.
Following shows a field mapping for a completion field named *title_suggest*:

```bash
PUT {INDEX_NAME}
{
 "mappings": {
  {TYPE_NAME}: {
   "properties": {
    "title_suggest": {
     "type": "completion"
   }
  }
 }
}
```
You can choose index and search time analyzer for the completion field by adding `analyzer` 
and `search_analyzer` options. 

### Context Mappings
Adding a `contexts` option in the field mapping defines a context-enabled completion field. You may want 
a context-enabled completion field, if you require filtering or boosting suggestions by a criteria other than 
just its prefix. Note that adding high-cardinality context values will increase the size of the in-memory 
index significantly.

There are two types of supported context types: `category` and `geo`.

#### Category Context Mapping
Category contexts are indexed as prefixes to the completion field value.

The following adds a category context named *genre*:
```bash
...
"contexts": [
 {
   "name": "genre",
   "type": "category"
 }
]
```
You can also pull context values from another field in a document by using a `path` option specifying the field name.

#### Geo Context Mapping
Geo points are encoded as geohash strings and prefixed to the completion field value.
The following adds a geo context named *location*:
```bash
...
"contexts": [
 {
   "name": "location",
   "type": "geo"
 }
]
```
You can also set `precision` option to choose the geohash length and `path` to pull context values from another
field in the document.

## Indexing
Just like any other field, you can add multiple completion fields to a document. You can also index multiple completions 
for a completion field per document. Each completion value is tied to its document and can be assigned an index-time
weight, which determines its relative rank among other completion values which share a common prefix.

The following indexes a completion value and its weight for the *title_suggest* completion field:
```bash
POST {INDEX_NAME}/{TYPE_NAME}
{
 "title_suggest": {
  "input": "title1",
  "weight": 7
 }
}
```
You can use the short-form, if you prefer not to add `weight` to the completions:
```bash
POST {INDEX_NAME}/{TYPE_NAME}
{
 "title_suggest": "title1",
}
```
Arrays are also supported to index multiple values,

The following indexes multiple completion entries (`input` and `weight`) for a single document:
```bash
POST {INDEX_NAME}/{TYPE_NAME}
{
 "title_suggest": [
  {
   "input": "title1",
   "weight": 14
  },
  {
   "input": "alternate_title",
   "weight": 7
  }
 ]
}
```
#### Indexing context-enabled fields
You can use the `path` option previously mentioned to pull context values from another field
in the document or add `contexts` option to the completion entry while indexing.

The following explicitly indexes context values along with completions:
```bash
POST {INDEX_NAME}/{TYPE_NAME}
{
 "genre_title_suggest": {
  "input": "title1",
  "contexts": {
   "genre": ["genre1", "genre2"]
  },
  "weight": 7
 }
}
```
You can also configure the `path` option in the context mapping to pull values from another 
field as follows (assuming `path` for the *genre* context has been set to *genre* field):
```bash
POST {INDEX_NAME}/{TYPE_NAME}
{
 "genre_title_suggest": "title1",
 "genre": ["genre1", "genre2"]
}
```

## Query Interface
The point of indexing values as completions is to be able to run fast prefix-based searches on them. 
You can run **Prefix**, **Fuzzy** and **Regex** queries on all completion fields. In case of a context-
enabled completion field, providing no context indicates  *all* contexts will be considered. But you
can not run a **Context** query on a completion field with no contexts. When a query is run on a context-
enabled field, the contexts for a completion is returned with the suggestion.

#### Prefix Query
The following suggests completions from the field *title_suggest* that start with *titl*:
```bash
POST {INDEX}/_suggest
{
 "suggest-namespace" : {
  "prefix" : "titl",
  "completion" : {
   "field" : "title_suggest"
  }
 }
}
```
The suggestions are sorted by their index-time `weight`.

#### Fuzzy Prefix Query
A fuzzy prefix query can serve typo-tolerant suggestions. It scores suggestions *closer* (based on its edit distance) 
to the provided *prefix* higher, regardless of their *weight*.
```bash
POST {INDEX}/_suggest
{
 "suggest-namespace" : {
  "prefix" : "sug",
  "completion" : {
   "field" : "suggest",
   "fuzzy" : {        (1)
    "fuzziness" : 2
   }
  }
 }
}
```
Specify *fuzzy* as shown in (1) to use typo-tolerant suggester. [Full options for fuzzy] (http://www.elastic.co/guide/en/elasticsearch/reference/current/search-suggesters-completion.html#fuzzy)

#### Regex Prefix Query

A regex prefix query matches all the term prefixes that match a regular expression. Regex is anchored at the begining but not at the end. 
The suggestions are sorted by their index-time `weight`.
```bash
POST {INDEX}/_suggest
{
 "suggest-namespace" : {
  "regex" : "s[u|a]g",    (1)
  "completion" : {
   "field" : "suggest"
  }
 }
}
```
Specify *regex* as shown in (1), instead of *prefix* to use regular expressions. [Supported regular expression syntax] (http://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-regexp-query.html#regexp-syntax)

### Context Query
Adding `contexts` (1) option to the query enables filtering and/or boosting suggestions based on their context values.
This query scores suggestions by multiplying the query-time *boost* withe the suggestion *weight*. 
```bash
POST {INDEX}/_suggest
{
 "suggest-namespace" : {
  "prefix" : "sug",
  "completion" : {
   "field" : "genre_title_suggest",
   "contexts": {           (1)
    "genre": [
     {
      "value" : "rock", 
      "boost" : 3
     },
     {
      "value" : "indie",
      "boost" : 2
     }
    ]
   }
  }
 }
}
```

The contexts can also be specified without any boost:

```bash
  ...
  "contexts": {
    "genre" : ["rock", "indie"]
  }
```
#### Geo Context Query:

The result will be scored such that the suggestions are first sorted by the distance between the corresponding geo context and the provided 
geo location and then by the weight of the suggestions.
```bash
  ...
  "contexts" : {
    "location" : {
      "context" : {
        "lat" : ..,
        "lon" : ..
      },
      "precision" : ..
    }
  }
```

#### Example
The following performs a **Fuzzy Prefix Query** combined with a **Context Query** on a context-enabled completion field named *genre_song_suggest*.

```bash
POST {INDEX}/_suggest
{
 "suggest-namespace" : {
  "prefix" : "like a roling st",
  "completion" : {
   "field" : "genre_song_suggest",
   "fuzzy" : {
    "fuzziness" : 2
   },
   "contexts" : {
    "genre": [
     {
      "context" : "rock", 
      "boost" : 3
     },
     {
      "context" : "indie",
      "boost" : 2
     }
    ]
   }
  }
 }
}
```
This query will return **all** song names for the genre *rock* and *indie* that are within an edit distance of *2* from the prefix *like a roling st*.
The song names with genre of *rock* will be boosted higher then that of *indie*.
The completion field values that share the longest prefix with *like a roling st* will be additionally boosted higher.

### Payload

You can retrieve any document field values along with its completions using the `payload` option.
The following returns the *url* field with each suggestion entry:
```bash
POST {INDEX}/_suggest
{
 "suggest-namespace" : {
  "prefix" : "titl",
  "completion" : {
   "field" : "title_suggest",
   "payload" : ["url"]
  }
 }
}
```
The response format is as follows:

```bash
{
 ...
 "suggest-namespace" : [ 
  {
   "prefix" : "sugg",
   "offset" : 0,
   "length" : 4,
   "options" : [ 
    {
     "text" : "suggestion",
     "score" : 34.0, 
     "payload": {
       "url" : [ "url_1" ]
     }
    }
   ]
  } 
 ]
}
```

### PR DESCRIPTION
**Pull request by areek**
Adds document-oriented completion suggester (see https://github.com/elastic/elasticsearch/issues/10746), replacing the previous completion suggester implementation from master.

This PR contains the changes from `feature/completion_suggester_v2` branch (https://github.com/elastic/elasticsearch/pull/11740 and https://github.com/elastic/elasticsearch/pull/13659).  Apart from code simplifications and bug fixes, the only functional change this adds is a collector which collects top N documents instead of top N suggestions. This was needed to ensure suggestion with multiple contexts would count as one suggestion. The change is contained in [CompletionSuggester.java](https://github.com/elastic/elasticsearch/compare/elastic:master...areek:completion_suggester_master?expand=1#diff-635ea967237e27a266c4b76ff1041016).

The [2.x branch](https://github.com/areek/elasticsearch/tree/completion_suggester_2x) and [2.1 branch](https://github.com/elastic/elasticsearch/tree/feature/completion_suggester_v2) for this feature implements the new completion suggester in a bwc way and allows the old suggester to be used if 
created in a previous version.

closes https://github.com/elastic/elasticsearch/issues/10746

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java [hunk 1]
```diff
@@ -18,97 +18,242 @@
...
+        int numResult = 0;
+        for (TopSuggestDocs.SuggestScoreDoc suggestScoreDoc : collector.get().scoreLookupDocs()) {
+            TopDocumentsCollector.SuggestDoc suggestDoc = (TopDocumentsCollector.SuggestDoc) suggestScoreDoc;
+            // collect contexts
+            Map<String, Set<CharSequence>> contexts = Collections.emptyMap();
+            if (fieldType.hasContextMappings() && !suggestDoc.getContexts().isEmpty()) {
+                contexts = fieldType.getContextMappings().getNamedContexts(suggestDoc.getContexts());
+            }
+            // collect payloads
+            Map<String, List<Object>> payload = Collections.emptyMap();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java]: can we just initialize this with `new HashMap<>()` and make it final?

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java [hunk 2]
```diff
@@ -18,97 +18,242 @@
...
 
     @Override
     protected Suggest.Suggestion<? extends Suggest.Suggestion.Entry<? extends Suggest.Suggestion.Entry.Option>> innerExecute(String name,
-            CompletionSuggestionContext suggestionContext, IndexSearcher searcher, CharsRefBuilder spare) throws IOException {
-        if (suggestionContext.fieldType() == null) {
-            throw new ElasticsearchException("Field [" + suggestionContext.getField() + "] is not a completion suggest field");
+            final CompletionSuggestionContext suggestionContext, final IndexSearcher searcher, CharsRefBuilder spare) throws IOException {
+        final CompletionFieldMapper.CompletionFieldType fieldType = suggestionContext.getFieldType();
+        if (fieldType == null) {
+            throw new ElasticsearchException("field [" + suggestionContext.getField() + "] is not a completion field");
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java]: can this be an "IllegalARgumentException"

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java [hunk 3]
```diff
@@ -30,13 +41,283 @@
...
+        /**
+         * Set to true if all measurements (like edit distance, transpositions and lengths) are in unicode
+         * code points (actual letters) instead of bytes. Default is false.
+         */
+        public FuzzyOptionsBuilder setUnicodeAware(boolean unicodeAware) {
+            this.unicodeAware = unicodeAware;
+            return this;
+        }
+
+        int getEditDistance() {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java]: can we get javadocs for the getters?

File: core/src/main/java/org/elasticsearch/search/suggest/completion/context/CategoryContextMapping.java [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+            queryContexts.add(innerParseQueryContext(parser));
+        } else if (token == Token.START_ARRAY) {
+            while (parser.nextToken() != Token.END_ARRAY) {
+                queryContexts.add(innerParseQueryContext(parser));
+            }
+        }
+        return queryContexts;
+    }
+
+    private CategoryQueryContext innerParseQueryContext(XContentParser parser) throws IOException, ElasticsearchParseException {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/context/CategoryContextMapping.java]: I wonder if you can use `ObjectParser` for this parsing? I can give you a hand if want with this?

File: core/src/main/java/org/elasticsearch/search/suggest/completion/context/CategoryQueryContext.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import java.io.IOException;
+
+import static org.elasticsearch.search.suggest.completion.context.CategoryContextMapping.CONTEXT_BOOST;
+import static org.elasticsearch.search.suggest.completion.context.CategoryContextMapping.CONTEXT_PREFIX;
+import static org.elasticsearch.search.suggest.completion.context.CategoryContextMapping.CONTEXT_VALUE;
+
+/**
+ * Defines the query context for {@link CategoryContextMapping}
+ */
+public class CategoryQueryContext implements ToXContent {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/context/CategoryQueryContext.java]: make this class final?

File: core/src/main/java/org/elasticsearch/search/suggest/completion/context/ContextMapping.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    /**
+     * Define a new context mapping of a specific type
+     *
+     * @param type type of context mapping, either {@link Type#CATEGORY} or {@link Type#GEO}
+     * @param name name of context mapping
+     */
+    protected ContextMapping(Type type, String name) {
+        super();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/context/ContextMapping.java]: super() is not needed

File: core/src/main/java/org/elasticsearch/search/suggest/completion/context/GeoContextMapping.java [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+            queryContexts.add(innerParseQueryContext(parser));
+        } else if (token == Token.START_ARRAY) {
+            while (parser.nextToken() != Token.END_ARRAY) {
+                queryContexts.add(innerParseQueryContext(parser));
+            }
+        }
+        return queryContexts;
+    }
+
+    private GeoQueryContext innerParseQueryContext(XContentParser parser) throws IOException, ElasticsearchParseException {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/context/GeoContextMapping.java]: I also wonder if we can use `ObjectParser` for this

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java [hunk 8]
```diff
@@ -18,97 +18,242 @@
...
+                } catch (CollectionTerminatedException e) {
+                    // collection was terminated prematurely
+                    // continue with the following leaf
+                }
+            }
+        }
+    }
+
+    // TODO: this should be refactored and moved to lucene
+    private static class TopDocumentsCollector extends TopSuggestDocsCollector {
```
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java]: I opened https://issues.apache.org/jira/browse/LUCENE-6880 for this

File: core/src/main/java/org/elasticsearch/search/suggest/completion/context/CategoryContextMapping.java (hunk 4 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/context/CategoryContextMapping.java]: Yes we can, I have taken a stab at this. `ObjectParser` is awesome, thanks for the pointer! There are a couple of failing tests for the `GeoContextMapping` after the switch, I will start investigating them tomorrow.

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java [hunk 9]
```diff
@@ -18,97 +18,241 @@
...
+                    scorer.score(collector.getLeafCollector(context), context.reader().getLiveDocs());
+                } catch (CollectionTerminatedException e) {
+                    // collection was terminated prematurely
+                    // continue with the following leaf
+                }
+            }
+        }
+    }
+
+    // TODO: this should be refactored and moved to lucene
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java]: This is https://issues.apache.org/jira/browse/LUCENE-6880 right?

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionFieldStats.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        if (fields != null  && fields.length > 0) {
+            completionFields = new ObjectLongHashMap<>(fields.length);
+        }
+        for (LeafReaderContext atomicReaderContext : indexReader.leaves()) {
+            LeafReader atomicReader = atomicReaderContext.reader();
+            try {
+                for (String fieldName : atomicReader.fields()) {
+                    Terms terms = atomicReader.fields().terms(fieldName);
+                    if (terms instanceof CompletionTerms) {
+                        // TODO: currently we load up the suggester for reporting it's size
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionFieldStats.java]: Does the current completion suggester do this too?

Also, it's -> its

**Comment by mikemccand** [comment]
LGTM, thanks @areek!

Nice to see we can now remove two forked Lucene classes, the custom postings format, and cutover to Lucene's impls instead.

**Comment by s1monw** [comment]
I left two minor comments on the latest commits, this LGTM too and doesn't need another review. Please go ahead and merge after the comments are fixed

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java (hunk 9 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java]: Yes, I added the issue url in the comment

File: core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionFieldStats.java (hunk 10 above)
Comment (areek) [inline core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionFieldStats.java]: fixed

**Comment by djschny** [comment]
Was this prematurely merged? I'm still pretty concerned about removing the payloads ability the users have now without concrete numbers that show the speed is the same. See comments on https://github.com/elastic/elasticsearch/issues/10746

**Comment by clintongormley** [comment]
This was originally merged into 2.2, but has been backed out in https://github.com/elastic/elasticsearch/pull/16170 and so is in master only.

**Pull request merged**